src/client.rs

```diff
@@ -250,21 +262,119 @@ impl<S: Read + Write> Client<S> {
+                    loop {
+                        raw_resp.clear();
+
+                        reader.read_line(&mut raw_resp)?;
+                        trace!("<== {}", raw_resp);
+
+                        let resp: serde_json::Value = serde_json::from_str(&raw_resp)?;
+
+                        /*match resp["method"].take().as_str() {
+                            Some(ref method) if method == &req.method => break 'outer resp,
+                            Some(ref method) => self.handle_notification(method, resp["result"].take())?,
+                            _ => break resp,
+                        };*/
+
+                        // Normally there is and id, but it's missing for spontaneous notifications
+                        // from the server
+                        let resp_id = resp["id"]
+                            .as_str()
+                            .and_then(|s| s.parse().ok())
+                            .or(resp["id"].as_u64().map(|i| i as usize));
+                        match resp_id {
+                            Some(resp_id) if resp_id == req.id => {
+                                // We have a valid id and it's exactly the one we were waiting for!
+                                trace!(
+                                    "Reader thread {} received a response for its request",
+                                    req.id
+                                );
+
+                                // Remove ourselves from the "waiting map"
+                                let mut map = self.waiting_map.write().unwrap();
+                                map.remove(&req.id);
+
+                                // If the map is not empty, we select a random thread to become the
+                                // new reader thread. We close its channel to "wake it up"
+                                if !map.is_empty() {
+                                    let key = *map.keys().nth(0).unwrap();
+
+                                    trace!(
+                                        "There are still some other threads waiting, waking up {}",
+                                        key
+                                    );
+                                    map.remove(&key);
+                                }
+
+                                break 'outer resp;
+                            }
+                            Some(resp_id) => {
+                                // We have an id, but it's not our response. Notify the thread and
+                                // move on
+                                trace!(
+                                    "Reader thread {}, received response for {}",
+                                    req.id,
+                                    resp_id
+                                );
+
+                                let mut map = self.waiting_map.write().unwrap();
+                                if map.get(&resp_id).is_some() {
```

Comment:
if let Some(sender) = map.remove(&resp_id) {
 sender.send(resp);
}

Reply:
It was like you proposed initially, but then I realized I was removing something from the map while i was also borrowing it from the map and I decided to change it.

Maybe it's actually possible to get it to work, I'll try